metadata/src/main/java/org/apache/kafka/timeline/timeline/SnapshottableHashTable.java

```diff
@@ -0,0 +1,410 @@
+            this.slot = 0;
+        }
+
+        @Override
+        public boolean hasNext() {
+            while (ready.isEmpty()) {
+                if (slot == topTier.length) {
+                    return false;
+                }
+                BaseHashTable.unpackSlot(temp, topTier, slot);
+                for (T object : temp) {
+                    if (object.startEpoch() <= snapshot.epoch()) {
+                        ready.add(object);
+                    }
+                }
+                temp.clear();
+
+                /*
+                 * As we iterate over the SnapshottableHashTable, elements may move from
+                 * the top tier into the snapshot tiers.  This would happen if something
+                 * were deleted in the top tier, for example, but still retained in the
+                 * snapshot.
+                 *
+                 * We don't want to return any elements twice, though.  Therefore, we
+                 * iterate over the top tier and the snapshot tier at the
+                 * same time.  The key to understanding how this works is realizing that
+                 * both hash tables use the same hash function, but simply choose a
+                 * different number of significant bits based on their size.
+                 * So if the top tier has size 4 and the snapshot tier has size 2, we have
+                 * the following mapping:
+                 *
+                 * Elements that would be in slot 0 or 1 in the top tier can only be in
+                 * slot 0 in the snapshot tier.
+                 * Elements that would be in slot 2 or 3 in the top tier can only be in
+                 * slot 1 in the snapshot tier.
+                 *
+                 * Therefore, we can do something like this:
+                 * 1. check slot 0 in the top tier and slot 0 in the snapshot tier.
+                 * 2. check slot 1 in the top tier and slot 0 in the snapshot tier.
+                 * 3. check slot 2 in the top tier and slot 1 in the snapshot tier.
+                 * 4. check slot 3 in the top tier and slot 1 in the snapshot tier.
+                 *
+                 * If elements move from the top tier to the snapshot tier, then
+                 * we'll still find them and report them exactly once.
+                 *
+                 * Note that while I used 4 and 2 as example sizes here, the same pattern
+                 * holds for different powers of two.  The "snapshot slot" of an element
+                 * will be the top few bits of the top tier slot of that element.
+                 */
+                HashTier<T> tier = snapshot.data(SnapshottableHashTable.this);
+                if (tier != null && tier.deltaTable != null) {
+                    BaseHashTable<T> deltaTable = tier.deltaTable;
+                    int shift = Integer.numberOfLeadingZeros(deltaTable.baseElements().length) -
+                        Integer.numberOfLeadingZeros(topTier.length);
```

Comment:
The top tier will always be at least as big as the lower tiers.  The reason is because an element only ends up in a lower tier if it first exists in the top tier and then is deleted.  Since the top tier never shrinks, only grows (similar to most hash tables), knowing that at least as many insertions have happened on the top tier as the lower tiers is sufficient to establish that it is always bigger.

If we ever implement shrinking we'll have to rethink this (but again, no mainstream Java implementation of hash tables does rehashing to shrink)

Reply:
I was thinking about the case for deletion. If we delete a bunch of keys in the top tier, all those previous values will be added to the snapshot tier. This can cause the hash table in the snapshot tier to be filled more than the one in the top tier. Since the resize is driven by the filling percentage, it seems that this could cause the snapshot tier to resize before the top tier does?

Reply:
If you have N items in the top tier, and you delete all of them, but they are retained by snapshot X, snapshot X now has N items.  But snapshot X cannot grow any more.  Any new deletions in the top tier won't end up in snapshot X because they will have happened after time X and therefore not be eligible.  So the limit on the size of snapshot X is N, which is the size of the top tier.  Since the expansion percentage is the same for both, they will end up with the same amount of slots, or the snapshot tier will have fewer.